File: server/schemas/query.py
"""Query schemas"""

from typing import Optional, List
from pydantic import BaseModel


class QueryGenerateRequest(BaseModel):
    """NL to SQL generation request."""
    user_query: str
    datasource_id: int


class QueryGenerateResponse(BaseModel):
    """NL to SQL generation response."""
    success: bool
    sql: str
    used_tables: List[str]
    error: Optional[str] = None


class QueryExecuteRequest(BaseModel):
    """SQL execution request."""
    sql: str
    datasource_id: int


class QueryExecuteResponse(BaseModel):
    """SQL execution response."""
    success: bool
    columns: List[str]
    rows: List[dict]
    row_count: int
    execution_time_ms: float
    error: Optional[str] = None


class QueryHistoryResponse(BaseModel):
    """Query history item."""
    id: int
    user_query: str
    generated_sql: str
    status: str
    execution_time_ms: Optional[float] = None
    row_count: Optional[int] = None
    used_tables: Optional[List[str]] = None
    created_at: str


class QueryFeedbackRequest(BaseModel):
    """Query feedback request."""
    query_log_id: int
    rating: Optional[int] = None
    is_correct: bool
    corrected_sql: Optional[str] = None
    feedback_text: Optional[str] = None


class ExportCreateRequest(BaseModel):
    """Export task creation request."""
    sql: str
    datasource_id: int


class ExportCreateResponse(BaseModel):
    """Export task creation response."""
    task_id: str
    status: str
    preview_rows: List[dict]
    preview_columns: List[str]
    preview_row_count: int
    total_rows: int
    message: str


class ExportStatusResponse(BaseModel):
    """Export task status response."""
    task_id: str
    status: str  # pending, processing, completed, failed
    total_rows: Optional[int] = None
    csv_content: Optional[str] = None
    error_message: Optional[str] = None
    message: str

File: client/src/api/query.ts
import axios from 'axios'

const request = axios.create({
  baseURL: '/api',
  timeout: 60000, // 60s for query execution
})

request.interceptors.request.use((config) => {
  const token = localStorage.getItem('token')
  if (token) {
    config.headers.Authorization = `Bearer ${token}`
  }
  return config
})

export interface GenerateSQLResponse {
  success: boolean
  sql: string
  used_tables: string[]
  error?: string
}

export interface ExecuteSQLResponse {
  success: boolean
  columns: string[]
  rows: Record<string, unknown>[]
  row_count: number
  execution_time_ms: number
  error?: string
}

export interface QueryHistoryItem {
  id: number
  user_query: string
  generated_sql: string
  status: string
  execution_time_ms?: number
  row_count?: number
  used_tables?: string[]
  created_at: string
}

export async function generateSQL(userQuery: string, datasourceId: number): Promise<GenerateSQLResponse> {
  const { data } = await request.post('/query/generate', {
    user_query: userQuery,
    datasource_id: datasourceId,
  })
  return data
}

export async function executeSQL(sql: string, datasourceId: number): Promise<ExecuteSQLResponse> {
  const { data } = await request.post('/query/execute', {
    sql,
    datasource_id: datasourceId,
  })
  return data
}

export async function getQueryHistory(limit = 50, offset = 0): Promise<QueryHistoryItem[]> {
  const { data } = await request.get('/query/history', {
    params: { limit, offset },
  })
  return data
}

export async function submitFeedback(
  queryLogId: number,
  isCorrect: boolean,
  rating?: number,
  correctedSql?: string
) {
  const { data } = await request.post('/query/feedback', {
    query_log_id: queryLogId,
    is_correct: isCorrect,
    rating,
    corrected_sql: correctedSql,
  })
  return data
}

// Export API
export interface ExportCreateResponse {
  task_id: string
  status: string
  preview_rows: Record<string, unknown>[]
  preview_columns: string[]
  preview_row_count: number
  total_rows: number
  message: string
}

export interface ExportStatusResponse {
  task_id: string
  status: 'pending' | 'processing' | 'completed' | 'failed'
  total_rows?: number
  csv_content?: string
  error_message?: string
  message: string
}

export async function createExportTask(sql: string, datasourceId: number): Promise<ExportCreateResponse> {
  const { data } = await request.post('/query/export', {
    sql,
    datasource_id: datasourceId,
  })
  return data
}

export async function getExportStatus(taskId: string): Promise<ExportStatusResponse> {
  const { data } = await request.get(`/query/export/${taskId}`)
  return data
}
